cypress-io/cypress
# feat: show running builds in the Debug page

- Closes https://github.com/cypress-io/cypress/issues/25759

### Additional details


Before this update, a developer would have to wait for a run to complete before being able to see the failed tests that may be a part of the run.  This change will allow the developer to switch to a running run and see any failed tests as they get recorded to the Cloud run.

#### Updates

* Updated `DebugNewRelevantRunBar`
  * Was moved above the Debug header when it is shown
  * Link for "View run" was changed to "Switch to run"
  * Link for "Switch to run" was added for a RUNNING run

_Before_
![image](https://user-images.githubusercontent.com/1702361/222204743-678e81ea-a7da-43f5-afd0-dd3771b4906a.png)
![image](https://user-images.githubusercontent.com/1702361/222204761-c7e348b6-1ee0-4f59-a463-cc3f4285773a.png)

_After_
<img width="1029" alt="Screenshot 2023-03-01 at 11 42 05 AM" src="https://user-images.githubusercontent.com/1702361/222205050-4b08f866-c42c-46d7-a30d-9669409de474.png">
<img width="1029" alt="Screenshot 2023-03-01 at 11 41 52 AM" src="https://user-images.githubusercontent.com/1702361/222205096-05e8175e-81f0-4b35-b4b2-718ba721cc94.png">

* Added `DebugTestingProgress` component
  * Shows when viewing a RUNNING run just below the header
  * Indicates how many specs have completed with text as well as a progress bar
  * Has a countdown when the run has a `completionScheduledAt` timestamp
  * Switches to say completion is soon if countdown expires

<img width="794" alt="Screenshot 2023-03-01 at 11 45 36 AM" src="https://user-images.githubusercontent.com/1702361/222206066-9a1af14a-df1f-47de-b9b6-88381be1ddec.png">
<img width="782" alt="Screenshot 2023-03-01 at 11 45 46 AM" src="https://user-images.githubusercontent.com/1702361/222206092-6715bf92-8f0e-42d6-bd02-5fae8afe05b5.png">
<img width="792" alt="Screenshot 2023-03-01 at 11 46 07 AM" src="https://user-images.githubusercontent.com/1702361/222206116-8050ad86-3884-438e-a692-5332da147a94.png">

_NOTE_ I have not implemented the "Run Completion Delay" tooltip and indication at the end of the progress bar.  I need to review with Product to see if we want to do this in a fast-follow

* Updated `DebugPendingRunSplash`
  * Removes the spec count because they are included in the `DebugTestingProgress` component
  * Update copy to support run is currently processing specs and has a state when the run is scheduled to complete

<img width="283" alt="Screenshot 2023-03-01 at 11 48 12 AM" src="https://user-images.githubusercontent.com/1702361/222207088-91101fae-0ccf-44b4-a2de-73ed2a8e8c3e.png">
<img width="346" alt="Screenshot 2023-03-01 at 11 48 27 AM" src="https://user-images.githubusercontent.com/1702361/222207113-e9e9e148-9f10-4cdc-bc65-5993f35ee9bf.png">

* Update `DebugPageHeader`
  * Allow for the duration field to count up for a running run

* Updated Sidebar badge to show when status of run is RUNNING

* Updates to support animating the changing parts of the UI
  * Updated the logic `Debug.vue` to cache and replace the data being passed to the `DebugContainer` to make sure the page does not flash when the query is rerunning to update the page
  * Added various `Transition` and `TransitionGroup` components to fade components in and out
  * Modified CSS and the structure of some components
    * Pulled internal padding from some components up to a parent container
    * Updated `StatsMetadata` to support a "prefix" slot so that the render in `DebugSpec` did not render a list (`ul`) within another list using incorrect DOM

* Backend updates
  * Updated `moveToNext` mutation to move to a specific run
  * Updates to `RelevantRunSpecsDataSource`
    * Track overall test counts to know when to trigger a running run to refresh
    * Track new cloud field called `scheduledToCompleteAt` to know when to switch the page to the "scheduled to complete" state
 
#### Extra credit
* Discovered a missing browser icon in `LayeredBrowserIcons` and added a placeholder.  Filed a ticket to get the real icon added.
* Update `useDurationFormat` to show milliseconds if the duration is less than 1 second

### Steps to test


* Component tests to review
  * `packages/app/src/debug/DebugContainer.cy.tsx`
    * Added "renders running run with failed tests" test to some what mimic the phases of a running run
      * NOTE: Does not mock out DebugTestingProgress component due to issues with different mocked Urql queries interfering with each other
  * `packages/app/src/debug/DebugNewRelevantRunBar.cy.tsx`
  * `packages/app/src/debug/DebugPageHeader.cy.tsx`
  * `packages/app/src/debug/DebugPendingRunSplash.cy.tsx`
  * `packages/app/src/debug/DebugTestingProgress.cy.tsx`
  * `packages/app/src/debug/StatsMetadata.cy.tsx`
  * `packages/app/src/navigation/SidebarNavigation.cy.tsx`

* E2E tests
  * The `debug.cy.ts` tests started to fail after including the `scheduledToCompleteAt` field in the query from `DebugContainer`.  When running locally, the page would thrash between states.  I believe the cause was the mocked up responses in `packages/app/cypress/fixtures/debug-Passing/gql-Debug.json` and `packages/app/cypress/fixtures/debug-Failing/gql-Debug.json` not containing this field.  Thus, Urql which see that it was not present and requery causing other queries to also rerun.  After adding the `scheduledToCompleteAt` field, the tests started passing again.
  * For this ticket, I relied on the CT tests for testing out the various states of the components and then live testing against a Cloud instance (staging and production).  I worked on adding a new E2E tests for the running state, but could not get it working in short order so left it out.  

* Options for real world testing
  * Setup a project on a local Cloud instance, on staging, or in production and manually record a run to it
  * Check out the branch for this PR and trigger a Circle CI run against it to watch it.  

### How has the user experience changed?


**See screenshots above**

### PR Tasks
